client/src/main/java/org/eclipse/hono/client/impl/DestinationCommandConsumer.java

```diff
@@ -0,0 +1,318 @@
+     * @throws IllegalArgumentException If the device or gateway id of the given handler does
+     *                                  not match this command consumer.
+     */
+    public Future<Void> addDeviceSpecificCommandHandler(final String deviceId, final String gatewayId,
+            final Handler<CommandContext> commandHandler, final Handler<Void> remoteCloseHandler) {
+        return addDeviceSpecificCommandHandler(new CommandHandlerWrapper(deviceId, gatewayId, commandHandler, remoteCloseHandler));
+    }
+
+    /**
+     * Adds a handler to this command consumer.
+     * 
+     * @param handler The consumer to add.
+     * @return A future indicating whether adding the handler was successful.
+     *         <p>
+     *         The future will be failed with a {@link ResourceConflictException}
+     *         if the consumer already contains a handler with the same device id.
+     *         <p>
+     *         Otherwise, the future will be succeeded.
+     * @throws NullPointerException If the given handler is {@code null}.
+     * @throws IllegalArgumentException If the device or gateway id of the given handler does
+     *                                  not match this command consumer.
+     */
+    public Future<Void> addDeviceSpecificCommandHandler(final CommandHandlerWrapper handler) {
+        Objects.requireNonNull(handler);
+        if (!handlerIsForConsumerGatewayOrDevice(handler.getDeviceId(), handler.getGatewayId())) {
+            LOG.error("cannot add handler with non-matching device/gateway id [consumer device id: {}, handler: {}",
+                    gatewayOrDeviceId, handler);
+            throw new IllegalArgumentException("invalid handler given");
+        }
+        if (commandHandlers.containsKey(handler.getDeviceId())) {
+            LOG.debug("cannot create concurrent command consumer [device-id: {}]", handler.getDeviceId());
+            return Future.failedFuture(new ResourceConflictException("message consumer already in use"));
+        }
+        commandHandlers.put(handler.getDeviceId(), handler);
+        return Future.succeededFuture(null);
+    }
+
+    private boolean handlerIsForConsumerGatewayOrDevice(final String handlerDeviceId, final String handlerGatewayId) {
+        return this.gatewayOrDeviceId.equals(handlerGatewayId)
+                || (handlerDeviceId.equals(this.gatewayOrDeviceId) && handlerGatewayId == null);
+    }
+
+    /**
+     * Checks whether the Proton receiver behind this consumer is open.
+     *
+     * @return {@code true} if this consumer is alive.
+     */
+    public boolean isAlive() {
+        return receiver.isOpen() && !closedCalled.get();
+    }
+
+    private void handleCommandMessage(final Message msg,  final ProtonDelivery delivery) {
+        // command could have been mapped to a gateway, but the original address stays in the message address in that case
+        final String originalDeviceId = msg.getAddress() != null
```

Comment:
Yes. The `DestinationCommandConsumer` is not only used for commands originating from another protocol adapter instance, but also for commands sent from applications that still use the legacy command & control link address.
Using that link address (`control/${tenant_id}/${device_id}`) directly from the application, there is no requirement to set the message `To` property.
I'll add a comment here.

Reply:
The C&C API does indeed require the address to be set on the message, though.
So, either the API is wrong (and we need to fix it) or this cannot happen here.

Reply:
The C&C API doc indeed already required the `To` address being set *before* the API changes for the new `command` endpoint. It looks to me this wasn't enforced in the code, though. So there could be applications using an empty address.
Since the legacy API will be removed in the next version anyway, I don't know whether we now need to add a check for a non-empty address in the code above. We could do that when removing support for the legacy control endpoint. WDYT?

Reply:
So, if this can only happen with the legacy endpoint, shouldn't we then be able to determine the device ID from the receiver link's source address?